File: Python/Medium/backtracking/word_search(2).py
class Solution:
    def exist(self, board, word):
        self.h = len(board)
        self.w = len(board[0])
        self.board = board

        for i in range(self.h):
            for j in range(self.w):
                if self.backtrack(i,j,word):
                    return True
        
        return False
    
    def backtrack(self,i,j,suffix):
        if len(suffix) == 0:
            return True
        
        if i<0 or i ==self.h or j <0 or j == self.w or self.board[i][j] != suffix[0]:
            return False
        
        ret = False
        self.board[i][j] = '#'
        directions = [(-1,0), (1,0), (0,-1), (0,1)]
        for dx, dy in directions:
            ret = self.backtrack(i+dx, j+dy, suffix[1:])
            if ret: break

        self.board[i][j] = suffix[0]

        return ret

File: JavaScript/Medium/backtracking/word_search(2).js
/**
 * @param {charachter [][]} board
 * @param {string} word
 * @return {boolean}
 */

var exist = function (board, word){
    let H = board.length
    let W = board[0].length

    var backtrack = (i,j,suffix) =>{
        if (suffix.length == 0){
            return true
        }

        if (i<0 || i==H || j <0 || j == W || board[i][j] != suffix[0]){
            return false
        }

        let ret = false
        board[i][j] = '#'

        let directions = [[-1,0], [1,0], [0,1], [0,-1]]
        for (let [dx,dy] of directions){
            ret = backtrack(i+dx,j+dy,suffix.slice(1))
            if (ret) {break}
        }

        board[i][j] = suffix[0]

        return ret
    }

    for (let i = 0; i<H; i++){
        for (let j =0; j<W; j++){
            if (backtrack(i,j,word)) {
                return true
            }
        }
    }
    return false

}